README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ghooks

<!-- badges: start -->

[![R-CMD-check](https://github.com/ThinkR-open/ghooks/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/ThinkR-open/ghooks/actions/workflows/R-CMD-check.yaml)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

The goal of `{ghooks}` is to provide `{ghooks}` hooks for standard
`{shiny}` apps.

## Installation

You can install the development version of `{ghooks}` like so:

``` r
pak::pak("thinkr-open/ghooks")
```

## About

You’re reading the doc about version : 0.0.0.9000

This README has been compiled on the

``` r
Sys.time()
[1] "2023-03-27 13:07:03 CEST"
```

Here are the test & coverage results :

``` r
devtools::check(quiet = TRUE)
ℹ Loading ghooks
── R CMD check results ────────────────────────────────── ghooks 0.0.0.9000 ────
Duration: 7.1s

0 errors ✔ | 0 warnings ✔ | 0 notes ✔
```

``` r
covr::package_coverage()
ghooks Coverage: 100.00%
R/detect_ui_line.R: 100.00%
R/geyser_hook.R: 100.00%
```

## `{golem}` hooks

`{golem}` hooks are functions that are called by `{golem}` at specific
moments of the app lifecycle. They are used to add custom code to the
app, notably when creating the app with `create_golem()`.

You can find more info about `{golem}` hooks on [`{golem}`
documentation](https://thinkr-open.github.io/golem/articles/f_extending_golem.html).

The hooks can be called with `golem::create_golem(project_hook = )`,
when creating a new golem project via the RStudio widget, as show on the
screenshot in the [Extending
`{golem}`](https://thinkr-open.github.io/golem/articles/f_extending_golem.html#defining-your-own-project_hook)
Vignette.

Below is the list of curently available hooks:

### `geyser_hook`

This hook is mainly intended at demoing the `{golem}` hooks system. It
adds the “Old Faithful Geyser Data” app to your golem project.

<details>
<summary>
Click here to see the content of `app_ui` and `app_server` from
`geyser_hook`
</summary>

``` r
pth <- tempfile(pattern = "geyser")
old_wd <- getwd()
unlink(pth, recursive = TRUE)
golem::create_golem(
  pth,
  project_hook = ghooks::geyser_hook,
  open = FALSE
)
── Checking package name ───────────────────────────────────────────────────────
✔ Valid package name
── Creating dir ────────────────────────────────────────────────────────────────
✔ Creating '/var/folders/9w/zdlv83ws6csdjnfc5x819gtr0000gn/T/RtmpMxqJJS/geyserd84144e998b8/'
✔ Setting active project to '/private/var/folders/9w/zdlv83ws6csdjnfc5x819gtr0000gn/T/RtmpMxqJJS/geyserd84144e998b8'
✔ Creating 'R/'
✔ Writing a sentinel file '.here'
• Build robust paths within your project via `here::here()`
• Learn more at <https://here.r-lib.org>
✔ Setting active project to '<no active project>'
✔ Created package directory
── Copying package skeleton ────────────────────────────────────────────────────
✔ Copied app skeleton
── Running project hook function ───────────────────────────────────────────────
Styling  1  files:
 R/app_ui.R ℹ 
────────────────────────────────────────
Status  Count   Legend 
✔   0   File unchanged.
ℹ   1   File changed.
✖   0   Styling threw an error.
────────────────────────────────────────
Please review the changes carefully!
Styling  1  files:
 R/app_server.R ℹ 
────────────────────────────────────────
Status  Count   Legend 
✔   0   File unchanged.
ℹ   1   File changed.
✖   0   Styling threw an error.
────────────────────────────────────────
Please review the changes carefully!
✔ All set
✔ Setting active project to
'/private/var/folders/9w/zdlv83ws6csdjnfc5x819gtr0000gn/T/RtmpMxqJJS/geyserd84144e998b8'
── Done ────────────────────────────────────────────────────────────────────────
A new golem named geyserd84144e998b8 was created at /var/folders/9w/zdlv83ws6csdjnfc5x819gtr0000gn/T//RtmpMxqJJS/geyserd84144e998b8 .
To continue working on your app, start editing the 01_start.R file.
```

``` r
cat(
  readLines(
    file.path(pth, "R", "app_ui.R")
  ),
  sep = "\n"
)
#' The application User-Interface
#'
#' @param request Internal parameter for `{shiny}`.
#'     DO NOT REMOVE.
#' @import shiny
#' @noRd
app_ui <- function(request) {
  tagList(
    # Leave this function for adding external resources
    golem_add_external_resources(),
    # Your application UI logic
    fluidPage(
      # Application title
      titlePanel("Old Faithful Geyser Data"),

      # Sidebar with a slider input for number of bins
      sidebarLayout(
        sidebarPanel(
          sliderInput(
            "bins",
            "Number of bins:",
            min = 1,
            max = 50,
            value = 30
          )
        ),

        # Show a plot of the generated distribution
        mainPanel(
          plotOutput("distPlot")
        )
      )
    )
  )
}

#' Add external Resources to the Application
#'
#' This function is internally used to add external
#' resources inside the Shiny application.
#'
#' @import shiny
#' @importFrom golem add_resource_path activate_js favicon bundle_resources
#' @noRd
golem_add_external_resources <- function() {
  add_resource_path(
    "www",
    app_sys("app/www")
  )

  tags$head(
    favicon(),
    bundle_resources(
      path = app_sys("app/www"),
      app_title = "geyserd84144e998b8"
    )
    # Add here other external resources
    # for example, you can add shinyalert::useShinyalert()
  )
}
```

``` r
cat(
  readLines(
    file.path(pth, "R", "app_server.R")
  ),
  sep = "\n"
)
#' The application server-side
#'
#' @param input,output,session Internal parameters for {shiny}.
#'     DO NOT REMOVE.
#' @import shiny
#' @noRd
app_server <- function(input, output, session) {
  # Your application server logic
  output$distPlot <- renderPlot({
    # generate bins based on input$bins from ui.R
    x <- faithful[, 2]
    bins <- seq(min(x), max(x), length.out = input$bins + 1)

    # draw the histogram with the specified number of bins
    hist(x, breaks = bins, col = "darkgray", border = "white")
  })
}
```

</details>

## About

You’re reading the doc about version : 0.0.0.9000

This README has been compiled on:

``` r
Sys.time()
[1] "2023-03-27 13:07:14 CEST"
```

Here are the test & coverage results :

``` r
devtools::check(quiet = TRUE)
ℹ Loading ghooks
── R CMD check results ────────────────────────────────── ghooks 0.0.0.9000 ────
Duration: 7.6s

0 errors ✔ | 0 warnings ✔ | 0 notes ✔
```

``` r
covr::package_coverage()
ghooks Coverage: 100.00%
R/detect_ui_line.R: 100.00%
R/geyser_hook.R: 100.00%
```


NEWS.md
# ghooks 0.0.0.9000 to 0.1.0

* Added the `geyser` hook

* Project Init

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(ghooks)

test_check("ghooks")


tests/testthat/helpers.R
if (!requireNamespace("golem", quietly = TRUE)) {
  install.packages("golem")
}
golem::install_dev_deps(force = TRUE)

tests/testthat/test-geyser_hook.R
# WARNING - Generated by {fusen} from /dev/flat_geyser.Rmd: do not edit by hand

test_that("geyser_hook works", {
  pth <- tempfile(pattern = "geyser")
  old_wd <- getwd()
  unlink(pth, recursive = TRUE)
  golem::create_golem(
    pth, project_hook = geyser_hook)
  expect_true(
    file.exists(
      file.path(pth, "R/app_ui.R")
    )
  )
  expect_true(
    file.exists(
      file.path(pth, "R/app_server.R")
    )
  )
  expect_true(
    grepl(
      "Old Faithful Geyser Data",
      paste(
        readLines(
          file.path(pth, "R/app_ui.R")
        ),
        collapse = ""
      )
    )
  )
  expect_true(
    grepl(
      "# Sidebar with a slider input for number of bins",
      paste(
        readLines(
          file.path(pth, "R/app_ui.R")
        ),
        collapse = ""
      )
    )
  )
  expect_true(
    grepl(
      "output\\$distPlot <- renderPlot",
      paste(
        readLines(file.path(pth, "R/app_server.R")),
        collapse = ""
      )
    )
  )
  expect_true(
    grepl(
      "histogram with the specified number of bins",
      paste(
        readLines(
          file.path(pth, "R/app_server.R")), collapse = "")
    )
  )
  setwd(old_wd)
})


R/shinydashboard_hook.R
#' shinydashboard hook
#'
#' This hook creates a shinydashboard applications taken from
#' https://rstudio.github.io/shinydashboard/get_started.html.
#' It is mainly inteded as a simple example of how to
#' use the `project_hook` argument of `create_golem()`.
#'
#' @return Used for side effect
#' @inheritParams golem::project_hook
#' @export
#'
#' @examples
#' if (FALSE){
#'   pth <- tempfile(pattern = "shinydashboard")
#'   golem::create_golem(pth, project_hook = shinydashboard_hook)
#'   readLines(file.path(pth, "R/app_ui.R"))
#'   unlink(pth, recursive = TRUE)
#' }
shinydashboard_hook <- function(path, package_name, ...) {
  withr::with_dir(path, {
    usethis::use_package("shinydashboard")
    shinydashboard_hook_replace_in_ui()
    shinydashboard_hook_replace_in_server()
  })
}

shinydashboard_hook_replace_in_ui <- function() {
  ui_line <- detect_ui_line()
  ui_lines <- readLines("R/app_ui.R")
  ui_lines[ui_line] <- '
     shinydashboard::dashboardPage(
      shinydashboard::dashboardHeader(title = "Basic dashboard"),
      shinydashboard::dashboardSidebar(),
      shinydashboard::dashboardBody(
        # Boxes need to be put in a row (or column)
        fluidRow(
          shinydashboard::box(plotOutput("plot1", height = 250)),

          shinydashboard::box(
            title = "Controls",
            sliderInput("slider", "Number of observations:", 1, 100, 50)
          )
        )
      )
    )'
  # replace fluidpage as well
  ui_lines <- ui_lines[-c(ui_line - 1, ui_line + 1)]
  writeLines(ui_lines, "R/app_ui.R")
  styler::style_file("R/app_ui.R")
}

shinydashboard_hook_replace_in_server <- function() {
  server_line <- detect_server_line()
  server_lines <- readLines("R/app_server.R")
  server_lines[server_line] <- '# Your application server logic
  set.seed(122)
  histdata <- rnorm(500)

  output$plot1 <- renderPlot({
    data <- histdata[seq_len(input$slider)]
    hist(data)
  })'
  writeLines(server_lines, "R/app_server.R")
  styler::style_file("R/app_server.R")
}


R/geyser_hook.R
# WARNING - Generated by {fusen} from /dev/flat_geyser.Rmd: do not edit by hand

#' geyser hook
#'
#' This hook creates an Old Faithful Geyser Data
#' application. It is mainly inteded as a simple example of how to
#' use the `project_hook` argument of `create_golem()`.
#'
#' @return Used for side effect
#' @inheritParams golem::project_hook
#' @export
#'
#' @examples
#' if (FALSE){
#'   pth <- tempfile(pattern = "geyser")
#'   golem::create_golem(pth, project_hook = geyser_hook)
#'   readLines(file.path(pth, "R/app_ui.R"))
#'   unlink(pth, recursive = TRUE)
#' }
geyser_hook <- function(path, package_name, ...) {
  withr::with_dir(path, {
    geyser_hook_replace_in_ui()
    geyser_hook_replace_in_server()
  })
}

geyser_hook_replace_in_ui <- function() {
  ui_line <- detect_ui_line()
  ui_lines <- readLines("R/app_ui.R")
  ui_lines[ui_line] <- '  # Application title
  titlePanel("Old Faithful Geyser Data"),

  # Sidebar with a slider input for number of bins
  sidebarLayout(
    sidebarPanel(
      sliderInput(
        "bins",
        "Number of bins:",
        min = 1,
        max = 50,
        value = 30
      )
    ),

    # Show a plot of the generated distribution
    mainPanel(
      plotOutput("distPlot")
    )
  )'
  writeLines(ui_lines, "R/app_ui.R")
  styler::style_file("R/app_ui.R")
}

geyser_hook_replace_in_server <- function() {
  server_line <- detect_server_line()
  server_lines <- readLines("R/app_server.R")
  server_lines[server_line] <- '# Your application server logic
  output$distPlot <- renderPlot({
    # generate bins based on input$bins from ui.R
    x <- faithful[, 2]
    bins <- seq(min(x), max(x), length.out = input$bins + 1)

    # draw the histogram with the specified number of bins
    hist(x, breaks = bins, col = "darkgray", border = "white")
  })'
  writeLines(server_lines, "R/app_server.R")
  styler::style_file("R/app_server.R")
}
